File: src/read-skeleton-38.js
// src/read-skeleton-38.js

/** 3.8.x equivalent of read-skeleton-42.js's walker. Field order verified
 *  against spine-ts/core/src/SkeletonBinary.ts @ 8b4844b (see README
 *  Provenance note) — independently, not assumed identical to 4.2.
 *  3.8 has no per-record flag byte anywhere: every field is read
 *  unconditionally except where marked nonessential-gated. */
export function walkHeaderAndConstraints38(r) {
  r.readFloat(); r.readFloat(); r.readFloat(); r.readFloat(); // x, y, width, height
  const nonessential = r.readBoolean();
  if (nonessential) {
    r.readFloat();  // fps
    r.readString(); // imagesPath
    r.readString(); // audioPath
  }

  const numStrings = r.readVarint(true);
  const strings = [];
  for (let i = 0; i < numStrings; i++) strings.push(r.readString());

  const numBones = r.readVarint(true);
  for (let i = 0; i < numBones; i++) {
    r.readString(); // name
    if (i !== 0) r.readVarint(true); // parent index
    r.readFloat(); r.readFloat(); r.readFloat(); // rotation, x, y
    r.readFloat(); r.readFloat(); // scaleX, scaleY
    r.readFloat(); r.readFloat(); // shearX, shearY
    r.readFloat(); // length
    r.readVarint(true); // transformMode (3.8: a varint enum index, not a raw byte)
    r.readBoolean(); // skinRequired
    if (nonessential) r.readInt32(); // color — 3.8 has no icon/visible fields
  }

  const numSlots = r.readVarint(true);
  for (let i = 0; i < numSlots; i++) {
    r.readString(); // name
    r.readVarint(true); // boneIndex
    r.readInt32(); // color
    r.readInt32(); // darkColor (always present, not nonessential-gated)
    r.readStringRef(strings); // attachmentName
    r.readVarint(true); // blendMode
  }

  skipIkConstraints38(r);
  skipTransformConstraints38(r);
  skipPathConstraints38(r);
  // No physics constraints in 3.8 — that kind is a 4.2-only addition.

  return { strings };
}

function skipIkConstraints38(r) {
  const n = r.readVarint(true);
  for (let i = 0; i < n; i++) {
    r.readString(); // name
    r.readVarint(true); // order
    r.readBoolean(); // skinRequired (a standalone boolean in 3.8, unlike 4.2's flag byte)
    const nn = r.readVarint(true);
    for (let j = 0; j < nn; j++) r.readVarint(true); // bone index
    r.readVarint(true); // target bone index
    r.readFloat(); // mix (unconditional)
    r.readFloat(); // softness (unconditional)
    r.readByte(); // bendDirection
    r.readBoolean(); // compress
    r.readBoolean(); // stretch
    r.readBoolean(); // uniform
  }
}

function skipTransformConstraints38(r) {
  const n = r.readVarint(true);
  for (let i = 0; i < n; i++) {
    r.readString(); // name
    r.readVarint(true); // order
    r.readBoolean(); // skinRequired
    const nn = r.readVarint(true);
    for (let j = 0; j < nn; j++) r.readVarint(true); // bone index
    r.readVarint(true); // target bone index
    r.readBoolean(); // local
    r.readBoolean(); // relative
    // 10 unconditional floats:
    r.readFloat(); // offsetRotation
    r.readFloat(); // offsetX
    r.readFloat(); // offsetY
    r.readFloat(); // offsetScaleX
    r.readFloat(); // offsetScaleY
    r.readFloat(); // offsetShearY
    r.readFloat(); // rotateMix
    r.readFloat(); // translateMix
    r.readFloat(); // scaleMix
    r.readFloat(); // shearMix
  }
}

function skipPathConstraints38(r) {
  const n = r.readVarint(true);
  for (let i = 0; i < n; i++) {
    r.readString(); // name
    r.readVarint(true); // order
    r.readBoolean(); // skinRequired
    const nn = r.readVarint(true);
    for (let j = 0; j < nn; j++) r.readVarint(true); // bone index
    r.readVarint(true); // target slot index
    r.readVarint(true); // positionMode (enum index)
    r.readVarint(true); // spacingMode (enum index)
    r.readVarint(true); // rotateMode (enum index)
    r.readFloat(); // offsetRotation
    r.readFloat(); // position
    r.readFloat(); // spacing
    r.readFloat(); // rotateMix
    r.readFloat(); // translateMix
  }
}

const ATTACHMENT_TYPES = ['Region', 'BoundingBox', 'Mesh', 'LinkedMesh', 'Path', 'Point', 'Clipping'];

export function readSkeleton38(r, nonessential) {
  const { strings } = walkHeaderAndConstraints38(r);
  const result = new Map();
  readSkin38(r, strings, nonessential, result);
  const otherSkinCount = r.readVarint(true);
  for (let i = 0; i < otherSkinCount; i++) {
    r.readStringRef(strings); // skin name (a table-index ref in 3.8, not an inline string)
    skipSkinBoneAndConstraintRefs38(r);
    readSkin38(r, strings, nonessential, null);
  }
  return result;
}

function readSkin38(r, strings, nonessential, outMap) {
  const slotCount = r.readVarint(true);
  for (let i = 0; i < slotCount; i++) {
    r.readVarint(true); // slotIndex
    const attachmentCount = r.readVarint(true);
    for (let ii = 0; ii < attachmentCount; ii++) {
      const attachmentName = r.readStringRef(strings);
      const info = readAttachment38(r, strings, attachmentName, nonessential);
      if (outMap && info) {
        const existing = outMap.get(info.path);
        // Deterministic tie-break on path collision: a Mesh conveys more
        // (uvs/triangles) than a Region, so it wins regardless of wire order.
        if (!existing || info.type === 'Mesh' || existing.type !== 'Mesh') {
          outMap.set(info.path, info);
        }
      }
    }
  }
}

function readAttachment38(r, strings, attachmentName, nonessential) {
  const name = r.readStringRef(strings) ?? attachmentName;
  const typeIndex = r.readUByte();
  const type = ATTACHMENT_TYPES[typeIndex];

  switch (type) {
    case 'Region': {
      const path = r.readStringRef(strings);
      r.readFloat(); // rotation
      r.readFloat(); // x
      r.readFloat(); // y
      r.readFloat(); // scaleX
      r.readFloat(); // scaleY
      r.readFloat(); // width
      r.readFloat(); // height
      r.readInt32(); // color, unconditional
      return { type: 'Region', path: path ?? name };
    }
    case 'BoundingBox': {
      const vertexCount = r.readVarint(true);
      readVertices38(r, vertexCount);
      if (nonessential) r.readInt32();
      return { type: 'BoundingBox', path: name };
    }
    case 'Mesh': {
      const path = r.readStringRef(strings);
      r.readInt32(); // color
      const vertexCount = r.readVarint(true);
      const uvs = readFloatArray38(r, vertexCount * 2);
      const triangleCount = r.readVarint(true);
      const triangles = [];
      for (let i = 0; i < triangleCount; i++) triangles.push(r.readShort());
      readVertices38(r, vertexCount);
      r.readVarint(true); // hullLength, unused by this library
      if (nonessential) {
        const edgeCount = r.readVarint(true);
        for (let i = 0; i < edgeCount; i++) r.readShort();
        r.readFloat(); // width
        r.readFloat(); // height
      }
      return { type: 'Mesh', path: path ?? name, uvs, triangles };
    }
    case 'LinkedMesh': {
      const path = r.readStringRef(strings);
      r.readInt32(); // color
      r.readStringRef(strings); // skinName
      r.readStringRef(strings); // parent
      r.readBoolean(); // inheritDeform (a real byte in 3.8, not a flag bit)
      if (nonessential) {
        r.readFloat(); // width
        r.readFloat(); // height
      }
      return { type: 'LinkedMesh', path: path ?? name };
    }
    case 'Path': {
      r.readBoolean(); // closed
      r.readBoolean(); // constantSpeed
      const vertexCount = r.readVarint(true);
      readVertices38(r, vertexCount);
      const lengthCount = vertexCount / 3;
      for (let i = 0; i < lengthCount; i++) r.readFloat();
      if (nonessential) r.readInt32();
      return { type: 'Path', path: name };
    }
    case 'Point': {
      r.readFloat(); // rotation
      r.readFloat(); // x
      r.readFloat(); // y
      if (nonessential) r.readInt32();
      return { type: 'Point', path: name };
    }
    case 'Clipping': {
      r.readVarint(true); // endSlotIndex
      const vertexCount = r.readVarint(true);
      readVertices38(r, vertexCount);
      if (nonessential) r.readInt32();
      return { type: 'Clipping', path: name };
    }
    default:
      throw new Error(`Unknown attachment type index: ${typeIndex}`);
  }
}

function readVertices38(r, vertexCount) {
  const floatCount = vertexCount * 2;
  const weighted = r.readBoolean();
  if (!weighted) {
    for (let i = 0; i < floatCount; i++) r.readFloat();
    return;
  }
  for (let i = 0; i < vertexCount; i++) {
    const boneCount = r.readVarint(true);
    for (let ii = 0; ii < boneCount; ii++) {
      r.readVarint(true); // bone index
      r.readFloat(); // x
      r.readFloat(); // y
      r.readFloat(); // weight
    }
  }
}

function readFloatArray38(r, n) {
  const arr = new Array(n);
  for (let i = 0; i < n; i++) arr[i] = r.readFloat();
  return arr;
}

function skipSkinBoneAndConstraintRefs38(r) {
  const boneCount = r.readVarint(true);
  for (let i = 0; i < boneCount; i++) r.readVarint(true); // bone index

  const ikCount = r.readVarint(true);
  for (let i = 0; i < ikCount; i++) r.readVarint(true); // IK constraint index

  const transformCount = r.readVarint(true);
  for (let i = 0; i < transformCount; i++) r.readVarint(true); // transform constraint index

  const pathCount = r.readVarint(true);
  for (let i = 0; i < pathCount; i++) r.readVarint(true); // path constraint index
}

File: skel_mesh_parser/read_skeleton_38.py
"""3.8.x skeleton binary walker -- port of
www/js/vendor/spine-skeleton-binary/read-skeleton-38.js. Field order matches
the vendored JS exactly (spec §6: mechanical port, no redesign). 3.8 has no
per-record flag byte anywhere: every field is read unconditionally except
where marked nonessential-gated."""
from __future__ import annotations

from typing import Any

from skel_mesh_parser.binary_reader import BinaryReader

_ATTACHMENT_TYPES = ["Region", "BoundingBox", "Mesh", "LinkedMesh", "Path", "Point", "Clipping"]


def walk_header_and_constraints_38(r: BinaryReader) -> tuple[list[str | None], bool]:
    r.read_float(); r.read_float(); r.read_float(); r.read_float()  # x, y, width, height
    nonessential = r.read_boolean()
    if nonessential:
        r.read_float()   # fps
        r.read_string()  # imagesPath
        r.read_string()  # audioPath

    num_strings = r.read_varint(True)
    strings: list[str | None] = [r.read_string() for _ in range(num_strings)]

    num_bones = r.read_varint(True)
    for i in range(num_bones):
        r.read_string()  # name
        if i != 0:
            r.read_varint(True)  # parent index
        r.read_float(); r.read_float(); r.read_float()  # rotation, x, y
        r.read_float(); r.read_float()  # scaleX, scaleY
        r.read_float(); r.read_float()  # shearX, shearY
        r.read_float()  # length
        r.read_varint(True)  # transformMode
        r.read_boolean()  # skinRequired
        if nonessential:
            r.read_int32()  # color

    num_slots = r.read_varint(True)
    for _ in range(num_slots):
        r.read_string()  # name
        r.read_varint(True)  # boneIndex
        r.read_int32()  # color
        r.read_int32()  # darkColor
        r.read_string_ref(strings)  # attachmentName
        r.read_varint(True)  # blendMode

    _skip_ik_constraints_38(r)
    _skip_transform_constraints_38(r)
    _skip_path_constraints_38(r)

    return strings, nonessential


def _skip_ik_constraints_38(r: BinaryReader) -> None:
    n = r.read_varint(True)
    for _ in range(n):
        r.read_string(); r.read_varint(True); r.read_boolean()
        nn = r.read_varint(True)
        for _ in range(nn):
            r.read_varint(True)
        r.read_varint(True)
        r.read_float(); r.read_float()
        r.read_byte(); r.read_boolean(); r.read_boolean(); r.read_boolean()


def _skip_transform_constraints_38(r: BinaryReader) -> None:
    n = r.read_varint(True)
    for _ in range(n):
        r.read_string(); r.read_varint(True); r.read_boolean()
        nn = r.read_varint(True)
        for _ in range(nn):
            r.read_varint(True)
        r.read_varint(True)
        r.read_boolean(); r.read_boolean()
        for _ in range(10):
            r.read_float()


def _skip_path_constraints_38(r: BinaryReader) -> None:
    n = r.read_varint(True)
    for _ in range(n):
        r.read_string(); r.read_varint(True); r.read_boolean()
        nn = r.read_varint(True)
        for _ in range(nn):
            r.read_varint(True)
        r.read_varint(True); r.read_varint(True); r.read_varint(True); r.read_varint(True)
        for _ in range(5):
            r.read_float()


def read_skeleton_38(r: BinaryReader) -> dict[str, dict[str, Any]]:
    strings, nonessential = walk_header_and_constraints_38(r)
    result: dict[str, dict[str, Any]] = {}
    _read_skin_38(r, strings, nonessential, result)
    other_skin_count = r.read_varint(True)
    for _ in range(other_skin_count):
        r.read_string_ref(strings)  # skin name (table-index ref in 3.8)
        _skip_skin_bone_and_constraint_refs_38(r)
        _read_skin_38(r, strings, nonessential, None)
    return result


def _read_skin_38(
    r: BinaryReader,
    strings: list[str | None],
    nonessential: bool,
    out_map: dict[str, dict[str, Any]] | None,
) -> None:
    slot_count = r.read_varint(True)
    for _ in range(slot_count):
        r.read_varint(True)  # slotIndex
        attachment_count = r.read_varint(True)
        for _ in range(attachment_count):
            attachment_name = r.read_string_ref(strings)
            info = _read_attachment_38(r, strings, attachment_name, nonessential)
            if out_map is not None and info is not None:
                existing = out_map.get(info["path"])
                if existing is None or info["type"] == "Mesh" or existing["type"] != "Mesh":
                    out_map[info["path"]] = info


def _read_attachment_38(
    r: BinaryReader,
    strings: list[str | None],
    attachment_name: str | None,
    nonessential: bool,
) -> dict[str, Any] | None:
    name = r.read_string_ref(strings)
    if name is None:
        name = attachment_name
    type_index = r.read_ubyte()
    attachment_type = _ATTACHMENT_TYPES[type_index]

    if attachment_type == "Region":
        path = r.read_string_ref(strings)
        r.read_float(); r.read_float(); r.read_float(); r.read_float()
        r.read_float(); r.read_float(); r.read_float()
        r.read_int32()
        return {"type": "Region", "path": path if path is not None else name}

    if attachment_type == "BoundingBox":
        vertex_count = r.read_varint(True)
        _read_vertices_38(r, vertex_count)
        if nonessential:
            r.read_int32()
        return {"type": "BoundingBox", "path": name}

    if attachment_type == "Mesh":
        path = r.read_string_ref(strings)
        r.read_int32()  # color
        vertex_count = r.read_varint(True)
        uvs = _read_float_array_38(r, vertex_count * 2)
        triangle_count = r.read_varint(True)
        triangles = [r.read_short() for _ in range(triangle_count)]
        _read_vertices_38(r, vertex_count)
        r.read_varint(True)  # hullLength, unused
        if nonessential:
            edge_count = r.read_varint(True)
            for _ in range(edge_count):
                r.read_short()
            r.read_float(); r.read_float()  # width, height
        return {"type": "Mesh", "path": path if path is not None else name, "uvs": uvs, "triangles": triangles}

    if attachment_type == "LinkedMesh":
        path = r.read_string_ref(strings)
        r.read_int32()  # color
        r.read_string_ref(strings)  # skinName
        r.read_string_ref(strings)  # parent
        r.read_boolean()  # inheritDeform
        if nonessential:
            r.read_float(); r.read_float()
        return {"type": "LinkedMesh", "path": path if path is not None else name}

    if attachment_type == "Path":
        r.read_boolean()  # closed
        r.read_boolean()  # constantSpeed
        vertex_count = r.read_varint(True)
        _read_vertices_38(r, vertex_count)
        # Real JS: `const lengthCount = vertexCount / 3;` (float division)
        # then `for (let i = 0; i < lengthCount; i++)` -- a `<` loop bound
        # against a possibly-fractional value iterates ceil(vertexCount/3)
        # times, NOT floor(vertexCount/3). Plain `//` here would silently
        # under-read one float for any vertexCount not a multiple of 3 (a
        # single open Bezier path segment is 4 control points, a real and
        # plausible shape, not an edge case) -- desyncing every subsequent
        # read in this skin with no exception raised. Found by Stage 2's
        # own scrutinize review; `(vertex_count + 2) // 3` is the integer
        # form of ceil(vertex_count / 3) for non-negative vertex_count.
        length_count = (vertex_count + 2) // 3
        for _ in range(length_count):
            r.read_float()
        if nonessential:
            r.read_int32()
        return {"type": "Path", "path": name}

    if attachment_type == "Point":
        r.read_float(); r.read_float(); r.read_float()
        if nonessential:
            r.read_int32()
        return {"type": "Point", "path": name}

    if attachment_type == "Clipping":
        r.read_varint(True)  # endSlotIndex
        vertex_count = r.read_varint(True)
        _read_vertices_38(r, vertex_count)
        if nonessential:
            r.read_int32()
        return {"type": "Clipping", "path": name}

    raise ValueError(f"Unknown attachment type index: {type_index}")


def _read_vertices_38(r: BinaryReader, vertex_count: int) -> None:
    float_count = vertex_count * 2
    weighted = r.read_boolean()
    if not weighted:
        for _ in range(float_count):
            r.read_float()
        return
    for _ in range(vertex_count):
        bone_count = r.read_varint(True)
        for _ in range(bone_count):
            r.read_varint(True)  # bone index
            r.read_float(); r.read_float(); r.read_float()  # x, y, weight


def _read_float_array_38(r: BinaryReader, n: int) -> list[float]:
    return [r.read_float() for _ in range(n)]


def _skip_skin_bone_and_constraint_refs_38(r: BinaryReader) -> None:
    bone_count = r.read_varint(True)
    for _ in range(bone_count):
        r.read_varint(True)
    ik_count = r.read_varint(True)
    for _ in range(ik_count):
        r.read_varint(True)
    transform_count = r.read_varint(True)
    for _ in range(transform_count):
        r.read_varint(True)
    path_count = r.read_varint(True)
    for _ in range(path_count):
        r.read_varint(True)
